File: protocol/src/frames.ts
/**
 * Tunnel frames: the application protocol carried inside the Noise channel.
 * One tunnel multiplexes every dsh interaction — unary RPC, the two event
 * downlinks, and cancellation — so the phone holds exactly one socket.
 *
 * Frames are JSON. The dsh API is JSON end to end (attachments ride as base64
 * inside it), so a binary framing would buy nothing and cost debuggability.
 */

/** Protocol version mixed into the Noise prologue; a mismatch aborts the handshake. */
export const TUNNEL_VERSION = 1

/**
 * Versions this build can speak, preferred first.
 *
 * A set rather than a number, because the two ends update independently: a
 * phone sits in the App Store review queue while the Bridle is a `npm
 * install` away, and after release version skew is the normal case rather
 * than the exception. Both ends offer what they can speak and the responder
 * picks the highest they share.
 */
/**
 * Largest single tunnel message any hop will carry, in bytes.
 *
 * A ceiling rather than a preference: every WebSocket implementation on the
 * path enforces one, and all of them enforce it the same brutal way — the
 * oversized message is never delivered and the *connection* is closed with a
 * 1009. There is no per-request failure and nothing to catch, so a sender that
 * does not check first produces a tunnel that drops, reconnects, resumes,
 * resends the same frame and drops again, with no layer able to say why.
 *
 * Which is why this is shared: both ends check against it before writing, so
 * the failure lands on the one request responsible instead of on the tunnel.
 *
 * 32 MiB is Cloudflare's limit for a message received by a Worker, and the
 * lowest ceiling on any path this protocol is deployed over. The Node relay
 * could carry more; matching the smallest one means a Bridle does not behave
 * differently depending on which relay it happens to be dialling.
 */
export const MAX_FRAME_BYTES = 32 * 1024 * 1024

export const TUNNEL_VERSIONS: readonly number[] = [1]

/**
 * Noise prologue both ends mix in before the first handshake message.
 *
 * Deliberately carries no version. An earlier design put one here, which made
 * a mismatch fail *inside* the handshake — before any channel exists, so the
 * refusal could not be sent and the client could not tell version skew from a
 * wrong machine key from tampering. The version is negotiated in the handshake
 * payload instead (see `HelloFrame` / `HandshakeReply`), which works because
 * the responder can always decrypt message one and can therefore always answer
 * with an *authenticated* refusal.
 */
export const TUNNEL_PROLOGUE: Buffer = Buffer.from('reins-tunnel', 'utf8')

/**
 * Choose the version two ends will speak.
 * @param offered - what the initiator says it supports, preferred first.
 * @param supported - what this build supports.
 * @returns the highest shared version, or undefined when there is no overlap.
 */
export function negotiateVersion(
  offered: readonly number[] | undefined,
  supported: readonly number[] = TUNNEL_VERSIONS,
): number | undefined {
  // An absent or empty list is the oldest client, which predates negotiation
  // and can only speak version 1.
  const wanted = offered === undefined || offered.length === 0 ? [1] : offered
  const shared = wanted.filter(version => supported.includes(version))
  return shared.length === 0 ? undefined : Math.max(...shared)
}

/** Which dsh downlink an event frame came from. */
export type StreamName = 'mux' | 'host'

/** App to Bridle: invoke one unary dsh method (`POST /api/<method>`). */
export interface RequestFrame {
  t: 'req'
  /** App-minted correlation id, unique per tunnel. */
  id: string
  /** dsh method path segment, e.g. `session.prompt` or `goals/create`. */
  method: string
  /** The dsh request payload (`{ args }` for Typert Remote methods). */
  payload: unknown
}

/** Bridle to App: the result of one {@link RequestFrame}. */
export interface ResponseFrame {
  t: 'res'
  id: string
  /** The dsh `server-response.result`, or a locally synthesized failure. */
  result: { ok: true; value: unknown } | { ok: false; error: { code: string; message: string; details: unknown } }
}

/** App to Bridle: abandon an in-flight request (maps to aborting the dsh fetch). */
export interface CancelFrame {
  t: 'cancel'
  id: string
}

/** App to Bridle: answer an approval or question (`POST /api/respond`). */
export interface RespondFrame {
  t: 'respond'
  id: string
  /** The dsh `client-response` message, verbatim. */
  message: unknown
}

/** Bridle to App: one downlink frame, tagged with a tunnel-level sequence. */
export interface EventFrame {
  t: 'ev'
  /** Monotonic per-tunnel sequence used by {@link ResumeFrame}. */
  seq: number
  stream: StreamName
  /** The dsh `server-request` frame, verbatim. */
  frame: unknown
}

/**
 * App to Bridle: where to knock when nobody is listening.
 *
 * The app posts its own notifications while it holds a tunnel, and that covers
 * the case where someone is already looking at their phone. It does not cover
 * the case this product exists for: the phone is in a pocket, iOS suspended the
 * process an hour ago, and the agent has just stopped to ask whether it may
 * delete something. Nothing local can fire, because nothing local is running.
 *
 * So the machine has to reach out, and the only thing that can wake a suspended
 * iOS app is APNs. The token travels inside the Noise channel — the Relay
 * carries it as ciphertext like everything else and learns it only at the
 * moment a push is actually sent, from the Bridle, one wake at a time.
 *
 * `token: null` means stop: someone turned notifications off in Settings, which
 * the app notices when it next comes to the foreground.
 *
 * No APNs environment travels with it. A token is minted against either the
 * development or the production host and is meaningless to the other, and an
 * earlier version had the app read `aps-environment` out of its own embedded
 * provisioning profile and pass the answer down through every layer. That is a
 * guess dressed as a fact — a Release build signed for development is a sandbox
 * token wearing a production badge — and Apple already answers the question by
 * refusing the wrong host with `BadDeviceToken`. The Relay tries both.
 */
export interface WakeFrame {
  t: 'wake'
  /** APNs device token, lowercase hex, or null to stop being woken. */
  token: string | null
}

/** App to Bridle: after a reconnect, replay everything past `since`. */
export interface ResumeFrame {
  t: 'resume'
  /** Highest sequence the app already has; `0` requests a fresh subscription. */
  since: number
}

/** Bridle to App: the replay buffer no longer reaches `since`; refetch state. */
export interface ResyncFrame {
  t: 'resync'
  /** First sequence the Bridle can still serve. */
  from: number
}

/** App to Bridle: opening frame stating what the app expects. */
export interface HelloFrame {
  t: 'hello'
  version: number
  /** Human-readable client build, shown in `bridle status`. */
  client: string
}

/** Bridle to App: connection is live; describes the machine and its dsh. */
export interface ReadyFrame {
  t: 'ready'
  version: number
  /** Bridle package version. */
  bridle: string
  /** Display name of the paired machine. */
  machine: string
  /** Whether the local dsh is reachable right now. */
  dshReachable: boolean
  /** dsh `host.describe` value when reachable. */
  host?: unknown
  /**
   * Where this machine can be dialled directly right now, best first.
   *
   * The pairing bundle's copy is frozen at pairing time; this one is current.
   * Absent from Bridles that predate it — a client must then keep whatever
   * addresses it has. Present-but-empty means the direct listener is off,
   * which is a reason to *drop* stored addresses, not keep them.
   */
  direct?: string[]
  /** Highest event sequence the Bridle has produced. */
  seq: number
}

/** Bridle to App: the local dsh went away or came back. */
export interface StatusFrame {
  t: 'status'
  dshReachable: boolean
  /** Operator-facing reason when unreachable. */
  detail?: string
}

/** Either direction: liveness probe. */
export interface PingFrame {
  t: 'ping'
  /** Echoed back verbatim. */
  nonce: string
}

/** Either direction: liveness answer. */
export interface PongFrame {
  t: 'pong'
  nonce: string
}

/** Bridle to App: a protocol-level refusal; the tunnel closes after it. */
export interface FaultFrame {
  t: 'fault'
  code: 'version' | 'unpaired' | 'internal' | 'busy'
  message: string
}

/** Frames the app may send. */
export type ClientFrame = HelloFrame | RequestFrame | CancelFrame | RespondFrame | ResumeFrame | WakeFrame | PingFrame | PongFrame

/** Frames the Bridle may send. */
export type ServerFrame = ReadyFrame | ResponseFrame | EventFrame | ResyncFrame | StatusFrame | PingFrame | PongFrame | FaultFrame

/** Every tunnel frame. */
export type TunnelFrame = ClientFrame | ServerFrame

/**
 * Encode one frame for the Noise channel.
 * @param frame - the frame to send.
 * @returns UTF-8 JSON bytes.
 */
export function encodeFrame(frame: TunnelFrame): Buffer {
  return Buffer.from(JSON.stringify(frame), 'utf8')
}

/**
 * Decode one frame received from the Noise channel.
 * @param bytes - the decrypted frame body.
 * @returns the parsed frame.
 * @throws {@link Error} when the body is not a JSON object carrying a string `t`.
 */
export function decodeFrame(bytes: Buffer): TunnelFrame {
  let parsed: unknown
  try {
    parsed = JSON.parse(bytes.toString('utf8'))
  } catch {
    throw new Error('tunnel frame is not JSON')
  }
  if (typeof parsed !== 'object' || parsed === null || typeof (parsed as { t?: unknown }).t !== 'string') {
    throw new Error('tunnel frame has no type tag')
  }
  return parsed as TunnelFrame
}

File: ios/Reins/Protocol/Frames.swift
/// Tunnel frames: the application protocol carried inside the Noise channel.
///
/// The Swift twin of `protocol/src/frames.ts`. One tunnel multiplexes everything
/// — unary RPC, both harness downlinks, cancellation — so the phone holds exactly
/// one socket no matter how much is going on.
///
/// Outbound frames state their key order (see `TunnelFrame`) rather than relying
/// on `JSONEncoder`, which has none to give.

import Foundation

/// Versions this build can speak, preferred first.
///
/// A set rather than a number, because the two ends update independently: this
/// app sits in a review queue while the Bridle is one `npm install` away, so
/// after release version skew is the normal case. Both ends offer what they can
/// speak and the machine picks the highest they share.
public let tunnelVersions: [Int] = [1]

/// The newest version this build speaks; what it reports about itself.
public let tunnelVersion = tunnelVersions[0]

/// Noise prologue both ends mix in before the first handshake message.
///
/// Deliberately carries no version. An earlier design put one here, which made
/// a mismatch fail *inside* the handshake — before any channel exists, so the
/// machine's refusal could not be sent and this end could not tell version skew
/// from a wrong machine key from tampering. The version is negotiated in the
/// handshake payload instead.
public let tunnelPrologue = Data("reins-tunnel".utf8)

/// Which harness downlink an event frame came from.
public enum StreamName: String, Codable, Sendable {
    /// Every session's events, aggregated.
    case mux
    /// Machine-level events: sessions added and removed, running flips, workspaces.
    case host
}

/// The encoder both ends agree on: no slash escaping, no pretty printing.
///
/// `JSONEncoder` escapes `/` by default and Node's `JSON.stringify` does not, so
/// without this a Typert Remote method like `goals/create` would go out as
/// `goals\/create` — still valid JSON, still routed correctly, but not the same
/// bytes, and the parity tests would be lying about agreement.
///
/// `sortedKeys` applies to the payloads nested inside a frame. A `JSONValue`
/// object is a Swift dictionary and has no order of its own, and Swift seeds its
/// string hashing per process, so without this the same payload serialises
/// differently between launches. Sorted is arbitrary but stable, which is what
/// makes a captured frame reproducible.
let tunnelEncoder: JSONEncoder = {
    let encoder = JSONEncoder()
    encoder.outputFormatting = [.withoutEscapingSlashes, .sortedKeys]
    return encoder
}()

/// A frame written with its keys in a stated order.
///
/// `JSONEncoder` backs a keyed container with a dictionary, so a `Codable`
/// struct's properties come out in whatever order hashing produced — not
/// declaration order, and not the same order twice. Node's `JSON.stringify`
/// emits insertion order, and the cross-implementation vectors compare bytes, so
/// the order has to be written down rather than inherited from a container that
/// does not have one.
///
/// Nothing here is signed, so a reordered frame would still be understood. The
/// point is that byte agreement is a property the two implementations can
/// actually be tested against, and a test that cannot fail is not a test.
protocol TunnelFrame: Sendable {
    /// The frame's members, in wire order. A `nil` value is omitted, matching
    /// `JSON.stringify` dropping `undefined`.
    var members: [(String, JSONValue?)] { get }
}

extension TunnelFrame {
    func encoded() throws -> Data { try orderedJSON(members) }
}

/// Serialise a JSON object with the given key order, dropping absent members.
func orderedJSON(_ members: [(String, JSONValue?)]) throws -> Data {
    var bytes = Data([UInt8(ascii: "{")])
    var first = true
    for (key, value) in members {
        guard let value else { continue }
        if !first { bytes.append(UInt8(ascii: ",")) }
        first = false
        bytes.append(try tunnelEncoder.encode(JSONValue.string(key)))
        bytes.append(UInt8(ascii: ":"))
        bytes.append(try tunnelEncoder.encode(value))
    }
    bytes.append(UInt8(ascii: "}"))
    return bytes
}

// MARK: - App to Bridle

/// Invoke one unary harness method (`POST /api/<method>` on the far side).
public struct RequestFrame: TunnelFrame {
    public let t = "req"
    /// App-minted correlation id, unique per tunnel.
    public let id: String
    /// Method path segment, e.g. `session.prompt` or `goals/create`.
    public let method: String
    /// The request payload (`{ args }` for Typert Remote methods).
    public let payload: JSONValue

    public init(id: String, method: String, payload: JSONValue) {
        self.id = id
        self.method = method
        self.payload = payload
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("id", .string(id)), ("method", .string(method)), ("payload", payload)]
    }
}

/// Abandon an in-flight request; the Bridle aborts its fetch to the harness.
public struct CancelFrame: TunnelFrame {
    public let t = "cancel"
    public let id: String

    public init(id: String) {
        self.id = id
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("id", .string(id))]
    }
}

/// Answer an approval or a question (`POST /api/respond` on the far side).
public struct RespondFrame: TunnelFrame {
    public let t = "respond"
    public let id: String
    /// The harness `client-response` message, verbatim.
    public let message: JSONValue

    public init(id: String, message: JSONValue) {
        self.id = id
        self.message = message
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("id", .string(id)), ("message", message)]
    }
}

/// After a reconnect, replay everything past `since`.
public struct ResumeFrame: TunnelFrame {
    public let t = "resume"
    /// Highest sequence the app already holds; `0` asks for a fresh subscription.
    public let since: Int

    public init(since: Int) {
        self.since = since
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("since", .number(Double(since)))]
    }
}

/// Liveness question, app to Bridle. The Bridle answers with a pong carrying
/// the same nonce — any frame at all proves the carrier, but asking forces an
/// answer out of a connection that would otherwise be silently dead for the
/// whole of the watchdog's patience.
public struct PingFrame: TunnelFrame {
    public let t = "ping"
    public let nonce: String

    public init(nonce: String) {
        self.nonce = nonce
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("nonce", .string(nonce))]
    }
}

/// Liveness answer.
public struct PongFrame: TunnelFrame {
    public let t = "pong"
    public let nonce: String

    public init(nonce: String) {
        self.nonce = nonce
    }

    var members: [(String, JSONValue?)] {
        [("t", .string(t)), ("nonce", .string(nonce))]
    }
}

/// What the app states inside handshake message one.
///
/// The token is present only while pairing. A device that is already known sends
/// none, so a stolen pairing QR is worth exactly one connection attempt.
public struct HandshakeRequest: TunnelFrame {
    /// Every version this build can speak, preferred first.
    public let versions: [Int] = tunnelVersions
    public let name: String
    public let client: String
    public let token: String?

    public init(name: String, client: String, token: String?) {
        self.name = name
        self.client = client
        self.token = token
    }

    var members: [(String, JSONValue?)] {
        [
            ("versions", .array(versions.map { .number(Double($0)) })),
            ("name", .string(name)),
            ("client", .string(client)),
            ("token", token.map(JSONValue.string)),
        ]
    }
}

/// What the Bridle states back inside handshake message two.
public struct HandshakeReply: Codable, Sendable {
    public let ok: Bool
    /// The version both ends will speak. Present when `ok`.
    public let version: Int?
    /// Refusal reason when `ok` is false: `version`, `unpaired`, or `internal`.
    public let reason: String?
    /// What the machine can speak, when it refused for `version`. Lets this end
    /// say *which* side is the old one rather than "something went wrong".
    public let supported: [Int]?
    /// Display name of the machine.
    public let machine: String?
    /// Bridle version string.
    public let bridle: String?

    /// Which end needs updating, for a refusal this end can explain.
    ///
    /// If the machine speaks something newer than anything we do, we are behind.
    /// Otherwise it is.
    public var weAreTheOldEnd: Bool {
        guard let supported, let theirBest = supported.max(), let ourBest = tunnelVersions.max() else { return false }
        return theirBest > ourBest
    }
}

// MARK: - Bridle to App

/// A harness call that failed, or a tunnel that could not carry it.
public struct CallError: Error, Equatable, Sendable {
    public let code: String
    public let message: String
    public let details: JSONValue

    public init(code: String, message: String, details: JSONValue = .null) {
        self.code = code
        self.message = message
        self.details = details
    }

    /// Whether retrying the same call could plausibly succeed.
    public var isTransient: Bool {
        ["disconnected", "timeout", "internal", "agent-busy"].contains(code)
    }
}

extension CallError: LocalizedError {
    public var errorDescription: String? { message }
}

/// Connection is live; describes the machine and its harness.
public struct ReadyFrame: Sendable {
    public let version: Int
    public let bridle: String
    public let machine: String
    public let dshReachable: Bool
    /// The harness `host.describe` value when reachable.
    public let host: JSONValue?
    /// Where this machine can be dialled directly right now, best first.
    ///
    /// `nil` from a Bridle too old to send it — which must leave the app's
    /// stored addresses alone. `[]` is different and deliberate: it means the
    /// direct listener is off, and keeping stale addresses around would have
    /// the app dialling a listener the operator turned off.
    public let direct: [String]?
    /// Highest event sequence the Bridle has produced.
    public let seq: Int
}

/// One downlink frame, tagged with a tunnel-level sequence.
public struct EventFrame: Sendable {
    public let seq: Int
    public let stream: StreamName
    /// The harness `server-request` frame, verbatim.
    public let frame: JSONValue
}

/// Everything the Bridle can send.
public enum ServerFrame: Sendable {
    case ready(ReadyFrame)
    case response(id: String, result: Result<JSONValue, CallError>)
    case event(EventFrame)
    /// The replay buffer no longer reaches back far enough; refetch state.
    case resync(from: Int)
    /// The local harness went away or came back.
    case status(reachable: Bool, detail: String?)
    case ping(nonce: String)
    case pong(nonce: String)
    /// A protocol-level refusal; the tunnel closes after it.
    case fault(code: String, message: String)
    /// A frame type this build does not know. Tolerated on purpose.
    case unknown(tag: String)
}

/// A frame that arrived malformed.
public struct FrameError: Error, LocalizedError, Equatable {
    public let reason: String
    public var errorDescription: String? { reason }
}

public extension ServerFrame {
    /// Parse one decrypted frame body.
    static func decode(_ bytes: Data) throws -> ServerFrame {
        guard let value = try? JSONValue(data: bytes) else { throw FrameError(reason: "tunnel frame is not JSON") }
        guard let tag = value["t"]?.stringValue else { throw FrameError(reason: "tunnel frame has no type tag") }
        switch tag {
        case "ready":
            return .ready(ReadyFrame(
                version: value["version"]?.intValue ?? tunnelVersion,
                bridle: value["bridle"]?.stringValue ?? "unknown",
                machine: value["machine"]?.stringValue ?? "a computer",
                dshReachable: value["dshReachable"]?.boolValue ?? false,
                host: value["host"],
                direct: value["direct"]?.arrayValue.map { $0.compactMap(\.stringValue) },
                seq: value["seq"]?.intValue ?? 0
            ))
        case "res":
            guard let id = value["id"]?.stringValue else { throw FrameError(reason: "response has no id") }
            let result = value["result"]
            if result?["ok"]?.boolValue == true {
                return .response(id: id, result: .success(result?["value"] ?? .null))
            }
            let error = result?["error"]
            return .response(id: id, result: .failure(CallError(
                code: error?["code"]?.stringValue ?? "internal",
                message: error?["message"]?.stringValue ?? "the machine reported a failure",
                details: error?["details"] ?? .null
            )))
        case "ev":
            guard let seq = value["seq"]?.intValue else { throw FrameError(reason: "event has no sequence") }
            let stream = StreamName(rawValue: value["stream"]?.stringValue ?? "mux") ?? .mux
            return .event(EventFrame(seq: seq, stream: stream, frame: value["frame"] ?? .null))
        case "resync":
            return .resync(from: value["from"]?.intValue ?? 0)
        case "status":
            return .status(reachable: value["dshReachable"]?.boolValue ?? false, detail: value["detail"]?.stringValue)
        case "ping":
            return .ping(nonce: value["nonce"]?.stringValue ?? "")
        case "pong":
            return .pong(nonce: value["nonce"]?.stringValue ?? "")
        case "fault":
            return .fault(
                code: value["code"]?.stringValue ?? "internal",
                message: value["message"]?.stringValue ?? "the machine refused the connection"
            )
        default:
            return .unknown(tag: tag)
        }
    }
}
